File: agentrocky/RockyState.swift
//
//  RockyState.swift
//  agentrocky
//

import SwiftUI
import Combine
import Darwin

enum RockyWall {
    case bottom
    case right
    case top
    case left
}

/// Shared observable state between AppDelegate (walk logic) and RockyView (display).
class RockyState: ObservableObject {
    @Published var walkFrameIndex: Int = 0
    @Published var jazzFrameIndex: Int = 0
    @Published var isJazzing: Bool = false
    @Published var direction: CGFloat = 1
    @Published var isChatOpen: Bool = false
    @Published var isDragging: Bool = false
    @Published var isAirborne: Bool = false
    @Published var isParachuteOpen: Bool = false
    @Published var isSleeping: Bool = false
    @Published var isPreparingJump: Bool = false
    @Published var isLookingAround: Bool = false
    @Published var landingPulse: Int = 0
    @Published var cornerGripPulse: Int = 0
    @Published var lookAroundPulse: Int = 0
    @Published var cursorProximity: CGFloat = 0
    @Published var cursorVectorX: CGFloat = 0
    @Published var cursorVectorY: CGFloat = 0
    @Published var cursorHoverPulse: Int = 0
    @Published var wall: RockyWall = .bottom
    @Published var positionX: CGFloat = 0
    @Published var positionY: CGFloat = 0
    var velocityX: CGFloat = 0
    var velocityY: CGFloat = 0
    var parachuteEligible: Bool = false
    @Published var speechBubble: String? = nil
    var clockwise: Bool = Bool.random()
    var screenBounds: CGRect = .zero
    var dockY: CGFloat = 0

    /// Single persistent Claude session — survives popover open/close.
    lazy var session: ClaudeSession = ClaudeSession(workingDirectory: realHome)

    private var realHome: String {
        getpwuid(getuid()).flatMap { String(cString: $0.pointee.pw_dir, encoding: .utf8) }
            ?? NSHomeDirectory()
    }
}

File: rocky_state.py
import os
import pwd
import random


class RockyWall:
    BOTTOM = 'bottom'
    TOP = 'top'
    LEFT = 'left'
    RIGHT = 'right'


class RockyState:
    def __init__(self):
        self.walk_frame_index = 0
        self.jazz_frame_index = 0
        self.is_jazzing = False
        self.direction = 1.0       # 1=face right, -1=face left
        self.is_chat_open = False
        self.is_dragging = False
        self.is_airborne = False
        self.is_parachute_open = False
        self.is_sleeping = False
        self.is_preparing_jump = False
        self.is_looking_around = False
        self.cursor_proximity = 0.0
        self.cursor_vector_x = 0.0
        self.cursor_vector_y = 0.0
        self.wall = RockyWall.BOTTOM
        self.pos_x = 0.0
        self.pos_y = 0.0           # Qt coords: top-left of window, y increases downward
        self.velocity_x = 0.0
        self.velocity_y = 0.0      # positive = downward (Qt)
        self.parachute_eligible = False
        self.speech_bubble = None
        self.clockwise = random.choice([True, False])
        # Screen bounds in Qt coords (set at startup)
        self.min_x = 0.0
        self.max_x = 0.0
        self.min_y = 0.0
        self.max_y = 0.0

    @property
    def real_home(self):
        try:
            return pwd.getpwuid(os.getuid()).pw_dir
        except Exception:
            return os.path.expanduser('~')
